persistence/sql/persister_consent.go

```diff
@@ -439,84 +439,103 @@ func (p *Persister) VerifyAndInvalidateLogoutRequest(ctx context.Context, verifi
 	/* #nosec G201 table is static */
 	var lr consent.LoginRequest
 	var lrh consent.HandledLoginRequest
 
 	var cr consent.ConsentRequest
 	var crh consent.HandledConsentRequest
 
-	return p.transaction(ctx, func(ctx context.Context, c *pop.Connection) error {
+	// The value of notAfter should be the minimum between input parameter and request max expire based on its configured age
+	requestMaxExpire := time.Now().Add(-p.config.ConsentRequestMaxAge())
+	if requestMaxExpire.Before(notAfter) {
+		notAfter = requestMaxExpire
+	}
 
-		// Delete all entries (and their FK)
-		// where hydra_oauth2_authentication_request were timed-out or rejected
-		// - hydra_oauth2_authentication_request_handled
-		// - hydra_oauth2_consent_request_handled
-		// AND
-		// - hydra_oauth2_authentication_request.requested_at < ttl.login_consent_request
-		// - hydra_oauth2_authentication_request.requested_at < notAfter
-
-		// Using NOT EXISTS instead of LEFT JOIN or NOT IN due to
-		// LEFT JOIN not supported by Postgres and NOT IN will have performance hits with large tables.
-		// https://stackoverflow.com/questions/19363481/select-rows-which-are-not-present-in-other-table/19364694#19364694
-		// Cannot use table aliasing in MYSQL, will work in Postgresql though...
-
-		err := p.Connection(ctx).RawQuery(fmt.Sprintf(`
-			DELETE
-			FROM %[1]s
-			WHERE NOT EXISTS
-				(
-				SELECT NULL
-				FROM %[2]s
-				WHERE %[1]s.challenge = %[2]s.challenge AND (%[2]s.error = '{}' OR %[2]s.error = '' OR %[2]s.error is NULL)
-				)
-			AND NOT EXISTS
-				(
-				SELECT NULL
-				FROM %[3]s
-				INNER JOIN %[4]s
-				ON %[3]s.challenge = %[4]s.challenge
-				WHERE %[1]s.challenge = %[3]s.login_challenge AND (%[4]s.error = '{}' OR %[4]s.error = '' OR %[4]s.error is NULL)
-				)
-			AND requested_at < ?
-			AND requested_at < ?
-			`,
-			(&lr).TableName(),
-			(&lrh).TableName(),
-			(&cr).TableName(),
-			(&crh).TableName()),
-			time.Now().Add(-p.config.ConsentRequestMaxAge()),
-			notAfter).Exec()
+	challenges := []string{}
+
+	// Select challenges from all consent requests that can be safely deleted with limit
+	// where hydra_oauth2_consent_request were unhandled or rejected, so either of these is true
+	// - hydra_oauth2_authentication_request_handled does not exist (unhandled)
+	// - hydra_oauth2_consent_request_handled has valid error (rejected)
+	// AND timed-out
+	// - hydra_oauth2_consent_request.requested_at < minimum between ttl.login_consent_request and notAfter
+	q := p.Connection(ctx).RawQuery(fmt.Sprintf(`
+		SELECT %[1]s.challenge
+		FROM %[1]s
+		LEFT JOIN %[2]s ON %[1]s.challenge = %[2]s.challenge
+		WHERE (
+			(%[2]s.challenge IS NULL)
+			OR (%[2]s.error IS NOT NULL AND %[2]s.error <> '{}' AND %[2]s.error <> '')
+		)
+		AND %[1]s.requested_at < ?
+		`,
+		(&cr).TableName(),
+		(&crh).TableName()),
+		notAfter)
+
+	if err := q.Limit(limit).All(&challenges); err == sql.ErrNoRows {
```

Comment:
If we limit we should probably order also?

Reply:
Idk, order operation is really heavy. Do you see any advantage in doing so? Is it worthy?

Reply:
If you order by PK it wouldn't be heavy! However, if this task is executed only in one worker, no order should be required. If it is executed in parallel, we need to order to avoid duplicate `DELETE` which might cause errors